javascript/ql/src/Performance/ReDoS.ql

```diff
@@ -693,6 +693,21 @@ predicate isFork(State q, InputSymbol s1, InputSymbol s2, State r1, State r2) {
     r1 != r2
     or
     r1 = r2 and q1 != q2
+    or
+    r1 = r2 and	
+    q1 = q2 and	
+    epsilonSucc+(q) = q and	
```

Comment:
I like your explanation -  I'm going to put it in. 

> The missing piece here is of course an explanation of why we check for a "mid" state being an infinite repetition quantifier as well. It seems that check is the reason we don't flag `(a?b?)*c` so I'm immediately a little suspicious of it.

That is correct. Without the `mid` check we would flag `(a?b?)*c`.  
But we would also flag `(a?)*`, which is a false positive (that only blows up quadraticly).   
So I want to keep it. 

> I believe the only correct fork state in (a?b?)*c is immediately after the a since there are two ways to read b afterwards: with or without the loop. The infinite repetition quantifier state is not actually a valid fork due to the restriction that a repetition body can't match the empty string, which means we can't use the epsilon-loop from that starting point.

I'm not sure what the right fork state is, but your proposal definitely sounds plausible.   
The problem is that there are lambda transitions from everything to everything, and it is therefore difficult to recognize the pattern based on the NFA.  

As I said in the previous PR. I think the best way to go is a more syntactic special case for the `(a?b?)*c` pattern. 
This would also allow make it easier to get the correct `pump` string.   
(Adding that special case is a thing for a later PR).

Reply:
A problem is that the fork for `(a?b?)*c` requires multiple transitions, and our `isFork` predicate only a single transition. 

The fork (as I see it) is between:   
1) `a` -> [skip `b` and go back to the start of the loop] -> [skip `a`] -> `b` 
2) `a` -> `b`

Both of these match the string `ab`, but they took 2 different routes in the NFA. 

I'm not sure it's feasible to expand `isFork` to support the multiple transition steps required for that fork.